Failed: akka.http.impl.engine.http2.Http2PersistentClientSpec.HTTP 2 managed persistent connection (tls = true) should transparently reconnect when connection is closed should when no requests are running

https://jenkins.akka.io:8498/view/Akka-http/job/akka-http-nightly/389/AKKA_VERSION=default,Node=akka-http-nightly-node,SCALA_VERSION=2.12.12,jdk=AdoptOpenJDK%208/testReport/junit/akka.http.impl.engine.http2/Http2PersistentClientSpec/HTTP_2_managed_persistent_connection__tls___true__should_transparently_reconnect_when_connection_is_closed_should_when_no_requests_are_running/

```
sbt.ForkMain$ForkError: java.lang.AssertionError: No severe log messages should be emitted during test run but got [1] warnings and [0] errors (see marked lines above)
	at akka.http.impl.util.WithLogCapturing.withFixture(WithLogCapturing.scala:65)
	at akka.http.impl.util.WithLogCapturing.withFixture$(WithLogCapturing.scala:23)
	at akka.http.impl.util.AkkaSpecWithMaterializer.withFixture(AkkaSpecWithMaterializer.scala:14)
```

caused by (rather harmless)

```
| [WARN] [01/13/2021 03:52:12.909] [Http2PersistentClientSpec-akka.actor.default-dispatcher-5] [akka://Http2PersistentClientSpec/system/IO-TCP/selectors/$a/6] unhandled message from Actor[akka://Http2PersistentClientSpec/deadLetters]: ChannelReadable
```

--- a/akka-http-core/src/test/scala/akka/http/impl/util/WithLogCapturing.scala
+++ b/akka-http-core/src/test/scala/akka/http/impl/util/WithLogCapturing.scala
@@ -18,7 +18,18 @@ import org.scalatest.{ Failed, Outcome, SuiteMixin, TestSuite }
 trait WithLogCapturing extends SuiteMixin { this: TestSuite =>
   implicit def system: ActorSystem
 
-  def failOnSevereMessages: Boolean = false
+  /**
+   * Can be overridden to return true to check that no warning or error messages are logged during
+   * the execution of the test
+   */
+  protected def failOnSevereMessages: Boolean = false
+
+  /**
+   * Can be overridden to adapt which events should be considered as severe if `failOnSevereMessages` is
+   * enabled.
+   */
+  protected def isSevere(event: LogEvent): Boolean =
+    event.level <= Logging.WarningLevel // yeah, lower levels are more severe
 
   abstract override def withFixture(test: NoArgTest): Outcome = {
     // When filtering just collects events into this var (yeah, it's a hack to do that in a filter).
@@ -47,6 +58,7 @@ trait WithLogCapturing extends SuiteMixin { this: TestSuite =>
         events.reverse.foreach { ev =>
           if (ev.level == WarningLevel) print(Console.YELLOW)
           else if (ev.level == ErrorLevel) print(Console.RED)
+          else if (isSevere(ev)) print(Console.YELLOW) // also mark other severe messages with YELLOW
           logger.print(ev)
           if (ev.level <= WarningLevel) print(Console.RESET)
         }
@@ -57,7 +69,7 @@ trait WithLogCapturing extends SuiteMixin { this: TestSuite =>
     if (!(res.isSucceeded || res.isPending)) {
       flushLog()
       res
-    } else if (failOnSevereMessages && events.exists(_.level <= Logging.WarningLevel)) {
+    } else if (failOnSevereMessages && events.exists(isSevere)) {
       val stats = events.groupBy(_.level).mapValues(_.size).toMap.withDefaultValue(0)
       flushLog()
 

--- a/akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2PersistentClientSpec.scala
+++ b/akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2PersistentClientSpec.scala
@@ -5,9 +5,10 @@
 package akka.http.impl.engine.http2
 
 import akka.actor.ActorSystem
+import akka.event.Logging
 import akka.http.impl.engine.ws.ByteStringSinkProbe
 import akka.http.impl.util.{ AkkaSpecWithMaterializer, ExampleHttpContexts }
-import akka.http.scaladsl.model.{ AttributeKey, AttributeKeys, ContentTypes, HttpEntity, HttpHeader, HttpMethod, HttpMethods, HttpRequest, HttpResponse, RequestResponseAssociation, StatusCode, StatusCodes, Uri, headers }
+import akka.http.scaladsl.model.{ AttributeKey, ContentTypes, HttpEntity, HttpHeader, HttpMethod, HttpMethods, HttpRequest, HttpResponse, RequestResponseAssociation, StatusCode, StatusCodes, Uri, headers }
 import akka.http.scaladsl.model.headers.HttpEncodings
 import akka.http.scaladsl.settings.ClientConnectionSettings
 import akka.http.scaladsl.unmarshalling.Unmarshal
@@ -35,6 +36,10 @@ class Http2PersistentClientSpec extends AkkaSpecWithMaterializer(
      akka.actor.serialize-messages = false
   """) with ScalaFutures {
   override def failOnSevereMessages: Boolean = true
+  override protected def isSevere(event: Logging.LogEvent): Boolean =
+    event.level <= Logging.WarningLevel &&
+      // fix for https://github.com/akka/akka-http/issues/3732 / https://github.com/akka/akka/issues/29330
+      !event.message.toString.contains("ChannelReadable")
 
   case class RequestId(id: String) extends RequestResponseAssociation
   val requestIdAttr = AttributeKey[RequestId]("requestId")

